### DISCUSSION TITLE
glb model is not visible

### DISCUSSION THREAD

**Discussion by bortomar-intellmaps** [visgl/loaders.gl#2876] [Q&A] [answered]
Hi, we are using @loaders.gl/gltf 3.4.6 and are experiencing a problem that we cannot identify the cause of. Some glb models are not showing up at all, while we are not getting any error and actually have no idea where we should look for the error message in this case. 
The screenshot below shows two models in the online glTF Viewer (https://gltf-viewer.donmccurdy.com) and the short clip should show both models in deck.gl MapView. 
The model on the right (in the browser dev tools it is BrainStem.glb) is displayed correctly in both instances, but the model on the left (stroj.glb) is only displayed in glTF Viewer.
Can you please advise?

![image](https://github.com/visgl/loaders.gl/assets/135999694/d0f3978d-1e98-4d03-bf1c-c4ba709d40ab)

[not-visible-glb.webm](https://github.com/visgl/loaders.gl/assets/135999694/0031ecab-00b7-4ba2-a593-6e4dee302d95)

**Comment by ibgreen** [comment]
Thanks fir reporting. Hard to say anything without testing the broken model.
- We don't support the two extensions used in the machine model, but I don't think that should prevent us from rendering as those are just optional advanced material properties that should be possible to ignore.
- You could put breakpoints in the luma.gl scenegraph ModelNode draw calls and see if it is trying to draw something.
- You already say that you don't see any console errors, so I won't ask you to check for that.

**Comment by ibgreen** [comment]
@donmccurdy

**Reply by donmccurdy** [reply]
Agreed with your summary above @ibgreen! The IOR and Specular extensions should not prevent displaying the model. It might be worth checking the scale of the model — sometimes models from different sources may be very large or small in world units. If the model uses skinning, that can also come with unexpected translation or scale factors. @bortomar-intellmaps if you can share a demo or the model in question, we may be able to debug further. Thanks!

**Reply by bortomar-intellmaps** [reply]
Hi, I don't know how to share demo but I attached a zip archive with the model. Thank you.

[stroj.zip](https://github.com/visgl/loaders.gl/files/14185373/stroj.zip)

**Reply by felixpalmer** [reply]
I don't have time right now to debug in detail, but I tried to load it using `examples/tutorials/hello-gltf/app.ts` in luma and I also don't see anything. It may be due to the camera being in the wrong place, the positions in the model are far from the origin
<img width="456" alt="Screenshot 2024-02-07 at 9 23 44" src="https://github.com/visgl/loaders.gl/assets/453755/42d4ec24-10fb-4f20-890a-911f81140645">

**Reply by donmccurdy** [reply, marked as answer]
As Felix suggests above, the cause of the issue is the large transform on the model. I processed the model to remove that transform by opening it in https://gltf.report/ and applying the script below:

```javascript
import { center, flatten, transformMesh } from '@gltf-transform/functions';

await document.transform(center(), flatten());

const scene = document.getRoot().getDefaultScene();

scene.traverse((node) => {
	const mesh = node.getMesh();
	if (!mesh) return;

	transformMesh(mesh, node.getMatrix(), true)

	node.setMatrix([
		1, 0, 0, 0,
		0, 1, 0, 0,
		0, 0, 1, 0,
		0, 0, 0, 1,
	]);
});
```

With that the transforms are removed and the model is centered in a box near the origin, with bounds <10 meters in on all sides, and the model appears as expected.

I see there is code in `examples/tutorials/hello-gltf/app.ts` intended to center the model regardless of its size ... it seems that isn't working quite correctly, or else it isn't intended to generalize to all models? Maybe we're using local AABBs where we need a global AABB or something? But in any case, the glTF rendering is working, the camera just isn't framing the scene.

**Comment by ibgreen** [comment]
Thanks @donmccurdy for the thorough investigation.
